internal/pipeline/pipeline_test.go

```diff
@@ -315,3 +317,266 @@ func isPipelineEqual(t *testing.T, p1, p2 Pipeline) bool {
+		if n.Valid && err != nil {
+			t.Fatalf("function name %s should be valid", n.Name)
+		}
+		if !n.Valid && err == nil {
+			t.Fatalf("function name %s should not be valid", n.Name)
+		}
+	}
+}
+
+func TestPipelineValidate(t *testing.T) {
+	type input struct {
+		Name  string
+		Input string
+		Valid bool
+	}
+	cases := []input{
+		{
+			Name: "no sources, no functions",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+`,
+			Valid: true,
+		},
+		{
+			Name: "have sources, no functions",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./base
+`,
+			Valid: true,
+		},
+		{
+			Name: "have sources and functions",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./base
+- ./*
+
+generators:
+- image: gcr.io/kpt-functions/generate-folders
+  config:
+    apiVersion: cft.dev/v1alpha1
+    kind: ResourceHierarchy
+    metadata:
+      name: root-hierarchy
+      namespace: hierarchy # {"$kpt-set":"namespace"}
+transformers:
+- image: patch-strategic-merge
+  configPath: ./patch.yaml
+- image: gcr.io/kpt-functions/set-annotation
+  configMap:
+    environment: dev
+
+validators:
+- image: gcr.io/kpt-functions/policy-controller-validate
+`,
+			Valid: true,
+		},
+		{
+			Name: "invalid apiversion",
+			Input: `
+apiVersion: kpt.dev/v1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./base
+`,
+			Valid: false,
+		},
+		{
+			Name: "absolute source path",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- /foo/bar
+`,
+			Valid: false,
+		},
+		{
+			Name: "invalid function name",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./*
+transformers:
+- image: patch@_@strategic-merge
+  configPath: ./patch.yaml
+`,
+			Valid: false,
+		},
+		{
+			Name: "more than 1 config",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./*
+transformers:
+- image: patch-strategic-merge
+  configPath: ./patch.yaml
+  configMap:
+    environment: dev
+`,
+			Valid: false,
+		},
+		{
+			Name: "absolute config path",
+			Input: `
+apiVersion: kpt.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: pipeline
+sources:
+- ./*
+transformers:
+- image: patch-strategic-merge
+  configPath: /patch.yaml
+`,
+			Valid: false,
+		},
+	}
+
+	for _, c := range cases {
+		b := bytes.NewBufferString(c.Input)
+		// FromReader will validate the pipeline
+		_, err := FromReader(b)
+		if c.Valid && err != nil {
+			t.Fatalf("%s: pipeline should be valid, %s", c.Name, err)
+		}
+		if !c.Valid && err == nil {
+			t.Fatalf("%s: pipeline should not be valid", c.Name)
+		}
+	}
+}
+
+func TestValidatePath(t *testing.T) {
+	type input struct {
+		Path  string
+		Valid bool
+	}
+
+	cases := []input{
+		{
+			Path:  "a/b/c",
+			Valid: true,
+		},
+		{
+			Path:  "a/b/",
```

Comment:
to reduce verbosity of table driven tests, you can omit field names Path and Valid